Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm

```diff
@@ -1333,4 +1333,100 @@ static void webProcessPoolHighDynamicRangeDidChangeCallback(CFNotificationCenter
 }
 #endif
 
+static AdditionalFonts additionalFonts(const Vector<URL>& fontURLs, std::optional<audit_token_t> auditToken)
+{
+    AdditionalFonts additionalFonts;
+    additionalFonts.fontDataList = WTF::compactMap(fontURLs, [auditToken = WTFMove(auditToken)](auto& fontURL) -> std::optional<FontData> {
+        std::optional<SandboxExtension::Handle> sandboxExtensionHandle;
+        if (auditToken)
+            sandboxExtensionHandle = SandboxExtension::createHandleForReadByAuditToken(fontURL.fileSystemPath(), *auditToken);
+        else
+            sandboxExtensionHandle = SandboxExtension::createHandle(fontURL.fileSystemPath(), SandboxExtension::Type::ReadOnly);
+
+        if (sandboxExtensionHandle)
+            return FontData { fontURL, WTFMove(*sandboxExtensionHandle) };
+        return FontData { fontURL, SandboxExtension::Handle() };
+    });
+    return additionalFonts;
+}
+
+void WebProcessPool::registerUserInstalledFonts(WebProcessProxy& process)
+{
+    if (m_userInstalledFontURLs) {
+        process.send(Messages::WebProcess::RegisterAdditionalFonts(additionalFonts(*m_userInstalledFontURLs, process.auditToken())), 0);
+        return;
+    }
+
+    auto blockPtr = makeBlockPtr([weakThis = WeakPtr { *this }] {
+        RetainPtr userInstalledFontsPath = [NSHomeDirectory() stringByAppendingPathComponent:@"Library/Fonts"];
+        RetainPtr enumerator = [NSFileManager.defaultManager enumeratorAtPath:userInstalledFontsPath.get()];
+
+        Vector<URL> fontURLs;
+        for (NSString *font in enumerator.get()) {
+            if ([font hasSuffix:@"ttf"] || [font hasSuffix:@"ttc"] || [font hasSuffix:@"otf"]) {
+                RetainPtr fontsPath = [userInstalledFontsPath stringByAppendingPathComponent:font];
+                auto fontURL = URL::fileURLWithFileSystemPath(String(fontsPath.get()));
+                fontURLs.append(fontURL);
+                RELEASE_LOG(Process, "Registering font url %s", fontURL.string().utf8().data());
+            }
+        }
+
+        RunLoop::protectedMain()->dispatch([weakThis = WTFMove(weakThis), fontURLs = crossThreadCopy(WTFMove(fontURLs))] {
+            if (RefPtr protectedThis = weakThis.get()) {
+                for (Ref process : protectedThis->m_processes)
+                    process->send(Messages::WebProcess::RegisterAdditionalFonts(additionalFonts(fontURLs, process->auditToken())), 0);
+                protectedThis->m_userInstalledFontURLs = WTFMove(fontURLs);
+            }
+        });
+    });
+
+    if (!m_hasStartedCachingUserInstalledFonts) {
+        dispatch_async(dispatch_get_global_queue(QOS_CLASS_DEFAULT, 0), blockPtr.get());
+        m_hasStartedCachingUserInstalledFonts = true;
+    }
+}
+
+static URL fontURLFromName(NSString *nsFontName)
+{
+    RetainPtr font = adoptCF(CTFontCreateWithName((__bridge CFStringRef)nsFontName, 0.0, nullptr));
```

Comment:
`ASCIILiteral` has `createCFString()`, so this could take `ASCIILiteral` directly